### PR TITLE
Implement padding and sanity checks for fused GEMM+GEMM 

### PR DESCRIPTION
**Pull request by rosenrodt**
Note: this does not yet follow the interface change in [conv_conv](https://github.com/ROCmSoftwarePlatform/composable_kernel/tree/conv_conv). FWIW this PR contains the necessary IsSupportedArgument() checks for passing padded/odd size unit tests. Hence the draft status.

Several fixes for `DeviceBatchedGemmGemm_Xdl_CShuffle`
1. Implement padding for GemmGemm's 4 dimensions MNKO
2. More unit tests for odd size & padded cases
3. Size requirements checks in `IsSupportedArgument()` necessary to pass unit tests
    - Check that matrix length in the contiguous dimension is integer multiple of SrcScalarPerVector 
    - Check that GEMM size K should be integer multiple of both AK1 and BK1

Generally speaking, item 3 isn't consistently implemented across many CK device ops. We may need to address that some time in the future.

### CODE REVIEW

**Comment by rosenrodt** [comment]
We need to decide what to do with this PR:
1. follow GemmSpec convention of using enum structs, API unchanged
2. follow what is implemented in conv_conv branch, API changed hence extra hours to get instances/tests sorted out

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp [hunk 1]
```diff
@@ -216,6 +217,13 @@ struct GridwiseBatchedGemmGemm_Xdl_CShuffle
             return false;
         }
 
+        // K is rounded to nearest multiples of K1 during tensor transformation so instead get KRaw
+        const auto KRaw = lengths_m_n_k_o[2];
+        if(!(KRaw % AK1 == 0 && KRaw % BK1 == 0))
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp:222]: should check the length with vector length, but AK1 and BK1 is not always the same as vector length

**Review by asroy — APPROVED** [review]

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp (hunk 1 above)
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp:222]: Is this what you refer to? Implemented in device level
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/tensor_operation/gpu/device/device_batched_gemm_gemm_xdl_cshuffle.hpp#L600-L616

Comment (asroy) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp:222]: Yes, so there is already this check. Great!

Why not put these 2 checks in the same place? Either in gridwise GEMM or DeviceOp?

Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp:222]: Ideally I want both checks in gridwise op. But the A/B/B1/C layout is lost in gridwise op (embedded in tensor descriptor). And inferring layout from tensor descriptor should work but isn't immediately obvious to me how. Suggestions?

Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp:222]: > Actually, why we need to check KRaw % AK1 == 0 && KRaw % BK1 == 0?

@asroy So I think about this again - this check is wrong and actually reject valid cases. Checking with ScalarPerVector is sufficient. I can remove it in next PRs. Although feel free to modify it while you are refactoring gridwise_batched_gemm_gemm

Comment (asroy) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp:222]: removed in PR #377

**Pull request merged**